test/mitmproxy/io/test_compat.py

```diff
@@ -24,3 +24,28 @@ def test_cannot_convert(tdata):
         flow_reader = io.FlowReader(f)
         with pytest.raises(exceptions.FlowReadException):
             list(flow_reader.stream())
+
+
+def test_convert_14_15(tdata):
```

Comment:
This test neatly obscures that `convert_14_15` is never called for the integration tests as you did not bump `mitmproxy.version.FLOWFORMAT_VERSION`. If you bump it to 15, the dumpfile-7-websocket test will fail because websocket messages are currently stored as tuples and not as dicts.
Long story short, I'm strongly in favor of removing the unit test here, we need to have full coverage by integration tests only.

Reply:
> we need to have full coverage by integration tests only.

Maybe I'm missing something, where are the actual integration tests? From what I see it looks like `test_compat.py` tests if the number of flows matches and that it doesn't raise. But I don't see tests for the actual logic of the migrations.

Reply:
Integration is indeed a bit far-fetched, I just adopted your terminology from https://github.com/mitmproxy/mitmproxy/pull/4691#issuecomment-883352513. 😉 I guess my pragmatic point is that they do actually raise here. 😄

Reply:
So how do I proceed? I think having unit tests for all migrations makes it much clearer what is happening. E.g. looking at something like `convert_9_10` it would help a lot to have a unit test so I can look at the before/after to understand what it does. Right now it appears that it's not tested at all what `convert_9_10` does, only that the migrations don't raise.

For `injected` in particular one issue is `dropped`, right? It was never migrated so that means the tuple is already missing something and I assume `dropped` will then be `None` after migrations? So maybe my migration should just push `False, False` at the end of the tuple?

Reply:
Following the tradition of `dropped` maybe I shouldn't add compat for `injected` at all and we separate the whole testing situation from this PR?